File: bgp-simulator.py
from optparse import OptionParser 
from random import * 
from Queue import Queue
from copy import deepcopy
#import numpy as np

IsDebug = True 
GraphEnabled = True 

AllNodeList = []
Adj_Matrix = [[]]
SDNList = []
NodeNum = 10
IsChanged = False
class path:
    def __init__(self, From):
        self.Validity = False
        self.Path = [From] 

class node:
    def __init__(self, ID):
        AllNodeList.append(self)
        self.ID = ID
        self.AdjList = []
        self.OldPathVector = []
        self.NewPathVector = []
        self.SDN = []
        for i in range(NodeNum):
            self.NewPathVector.append(path(ID))
        
    def add_adj(self, ID):
        self.AdjList.append(ID) 
        if self.NewPathVector[ID].Validity == False:
            self.NewPathVector[ID].Validity = True
            self.NewPathVector[ID].Path.append(ID)
        else:
            self.NewPathVector[ID].Path = [self.ID,ID]
        #myprint_path(0) 

    def bgp_update(self):
        global IsChanged
        for i in range(NodeNum):
            ChangedInThisLoop = False
            if i in self.AdjList:#avoid handling direct neighbor, waste time
                continue
            if self.OldPathVector[i].Validity == False:
                TmpPathLength = float('inf')
            else:
                TmpPathLength = len(self.OldPathVector[i].Path)
            for Adj in self.AdjList:
                if AllNodeList[Adj].OldPathVector[i].Validity == False:
                    continue
                if self.ID in AllNodeList[Adj].OldPathVector[i].Path:#avoid loop 
                    continue 
                if len(AllNodeList[Adj].OldPathVector[i].Path) + 1 < TmpPathLength:
                    TmpPathLength = len(AllNodeList[Adj].OldPathVector[i].Path)+1
                    MarkedID = Adj
                    ChangedInThisLoop = True 
            if ChangedInThisLoop == True: 
                    self.NewPathVector[i].Validity = True
                    self.NewPathVector[i].Path = deepcopy(AllNodeList[MarkedID].OldPathVector[i].Path)
                    self.NewPathVector[i].Path.insert(0,self.ID)
                    IsChanged = True

class SDN:
    def __init__(self, SdnID, SdnSize):
        if SdnSize == 0:
            return
        self.ID = SdnID
        self.Size = 0
        self.AsList = []
        SDNList.append(self)
        self.create(SdnSize)
    def create(self, SdnSize): #keep adding new ASes
        MyQueue = Queue(-1)
        StartNodeID = int(random()*NodeNum)
        print "# R# StartID: ", StartNodeID
        self.add_AS(StartNodeID)
        MyQueue.put(StartNodeID)
        while self.Size <> SdnSize: 
            CurrentNodeID = MyQueue.get() 
            for AdjNodeID in AllNodeList[CurrentNodeID].AdjList:
                if self.ID in AllNodeList[AdjNodeID].SDN:
                    continue
                self.add_AS(AdjNodeID)
                MyQueue.put(AdjNodeID)
                if self.Size == SdnSize:
                    break
    def add_AS(self, AsID):
        self.Size = self.Size + 1
        self.AsList.append(AsID)
        AllNodeList[AsID].SDN.append(self.ID)
    def local_converge(self):
        global IsChanged
        TmpIsChanged = IsChanged
        while True:
            #myprint_path()
            IsChanged = False
            for i in self.AsList:
                AllNodeList[i].OldPathVector = deepcopy(AllNodeList[i].NewPathVector)
            for i in self.AsList:
                AllNodeList[i].bgp_update()
            if IsChanged == False:
                break
        IsChanged = TmpIsChanged #recover info about whether modified in normal update (non-SDN), SDN converge cannot change IsChanged
        #myprint_path()
        
def myprint_matrix(M):
    if not IsDebug:
        return
    for i in range(len(M)):
        print '(%d) ' %i,
        for j in range(len(M[i])):
            print M[i][j],
        print ''
        
def myprint_path(ID=-1):
    if not IsDebug:
        return
    if not GraphEnabled:
        return 
    print ''
    print 'begin to print path:'
    if ID == -1:
        print '============================'
        for ID in range(NodeNum):
            print '%d ------' %ID
            for i in range(len(AllNodeList[ID].NewPathVector)):
                print '(%d): ' %i,
                if AllNodeList[ID].NewPathVector[i].Validity == False:
                    print "Invalidate"
                else:
                    for j in range(len(AllNodeList[ID].NewPathVector[i].Path)):
                        print '%3d ' %AllNodeList[ID].NewPathVector[i].Path[j],
                    print ''
    else:
        print "below print the path in node %d :" % ID
        for i in range(len(AllNodeList[ID].NewPathVector)):
            print '(%d): ' %i,
            if AllNodeList[ID].NewPathVector[i].Validity == False:
                print "Invalidate"
                for j in range(len(AllNodeList[ID].NewPathVector[i].Path)):
                    print '%3d ' %AllNodeList[ID].NewPathVector[i].Path[j],
                print ''

def myprint_graph(adjacency_matrix, name):
    if not IsDebug:
        return
    if not GraphEnabled:
        return 
    import networkx as nx
    import matplotlib.pyplot as plt

    adjacency_matrix = np.array(adjacency_matrix)
    rows, cols = np.where(adjacency_matrix == 1)
    edges = zip(rows.tolist(), cols.tolist())
    gr = nx.Graph()
    gr.add_edges_from(edges)
    # nx.draw(gr) # edited to include labels
    nx.draw_networkx(gr)
    # now if you decide you don't want labels because your graph
    # is too busy just do: nx.draw_networkx(G,with_labels=False)
    #plt.show()
    plt.savefig('topo_'+name+'.pdf')

def myprint_graphfile(adjacency_matrix, name):
    if not IsDebug:
        return
    if not GraphEnabled:
        return
      

def SFENG(Nodes, mlinks, seed):
    Net = [[0 for col in range(Nodes)] for row in range(Nodes)]
    pos = len(seed)
    for i in range(pos):
        for j in range(pos):
            Net[i][j] = seed[i][j]
    sumlinks = 0
    for i in range(pos):
        sumlinks = sumlinks + sum(Net[i])
    while pos < Nodes:
        linkage = 0
        if mlinks == -1:
            tmp = random()
            if tmp < 0.6:#override mlinks.
                mlinks = 1
            elif tmp < 0.9:
                mlinks = 2
            else:
                mlinks = 3
        while linkage <> mlinks:
            rnode = int(random()*pos)
            deg = sum(Net[rnode]) * 2
            rlink = random() 
            if rlink < float(deg) / sumlinks and Net[pos][rnode] <> 1 and Net[rnode][pos] <> 1:
                Net[pos][rnode] = 1
                Net[rnode][pos] = 1
                linkage = linkage + 1
                sumlinks = sumlinks + 2
        pos = pos + 1
    return Net

def converge_process():
    global IsChanged
    CurrentNodeNum = len(AllNodeList)
    LoopTimes = 0
    while True:
        #myprint_path()
        IsChanged = False
        for i in range(CurrentNodeNum):
            AllNodeList[i].OldPathVector = deepcopy(AllNodeList[i].NewPathVector)
        for i in range(CurrentNodeNum):
            AllNodeList[i].bgp_update()
        for sdn in SDNList:
            sdn.local_converge()
        LoopTimes = LoopTimes + 1
        if IsChanged == False:
            break
    myprint_path()
    return LoopTimes

def inject_up(i):
    # currently havenot consider the locality of the up link, ID1 and ID2 may not be close to each other
    CurrentNode = node(i)
    for j in range(i):
        if Adj_Matrix[i][j]:
            CurrentNode.add_adj(j)
            AllNodeList[j].add_adj(i)
            print  "#   R# link up info injected", i, j 
            break


if __name__ == "__main__":
    Parser = OptionParser()
    Parser.add_option("-n","--nodenum",dest="NodeNum",default=20)
    Parser.add_option("-r","--sdnratio",dest="SdnRatio",default=20)#The unit of SdnRatio is percent
    Parser.add_option("-s","--seed",dest="Seed",default=0)#The unit of SdnRatio is percent
    (options, args) = Parser.parse_args()
    NodeNum = int(options.NodeNum)
    SdnRatio = float(options.SdnRatio)/100.0

    seed(int(options.Seed))
    
    myseed =[[0,1,0,0,1],[1,0,0,1,0],[0,0,0,1,0],[0,1,1,0,0],[1,0,0,0,0]] 
    Adj_Matrix = SFENG(NodeNum, -1, myseed)
    myprint_matrix(Adj_Matrix)
    myprint_graph(Adj_Matrix,'0')
    
    for i in range(NodeNum-10):
        CurrentNode = node(i)
        for j in range(NodeNum-10):
            if Adj_Matrix[i][j]:
                CurrentNode.add_adj(j)
    #print AllNodeList
    print "initial converge time: # %d" % converge_process()
    SDN(0,int(NodeNum*SdnRatio))
    for i in range(10):
        inject_up(NodeNum - 10 + i)
        print "update converge time: # %d" %converge_process()
        print "# Update #%d over" %i

File: bgp-simulator.cpp
#include<stdio.h>
#include<stdlib.h>
#include<memory.h>
#include<vector>
#include<list>
#include<string>
#include<algorithm>

using namespace std;
typedef unsigned short u_s;

//#define DEBUG

class node;
class SDN;

u_s NodeNum = 0;
u_s ValidNodeNum = 0;
bool IsChanged = 0;
u_s RepeatTimes = -1;
int RandSeed = 0;
float SDNRatio = -0.1;


SDN * MySDN;
vector<node *> AllNodeVector;
bool ** AdjMatrix;

void myprint_path();
u_s converge_process(bool sign);
void SFENG(bool ** AdjMatrix, bool MySeed[5][5]);
void inject_up(u_s );

class path{
	public:
		bool Validity;
		list<u_s> Path;
		path(u_s From){
			Validity = 0;
			Path.push_back(From);
		}
};	
class node{
	public:
		u_s ID;
		list<u_s> AdjList;
		vector<path> OldPathVector;
		vector<path> NewPathVector;
		bool IsInSDN;
		node(u_s myID): OldPathVector(NodeNum, path(myID)), NewPathVector(NodeNum, path(myID)){
			this->ID = myID;
			this->IsInSDN = 0;
		}
		void add_adj(u_s myID){
			AdjList.push_back(myID);
			if(!NewPathVector[myID].Validity){
				NewPathVector[myID].Validity = 1;
				NewPathVector[myID].Path.push_back(myID);
			}
			else{
				NewPathVector[myID].Path.clear();
				NewPathVector[myID].Path.push_back(ID);
				NewPathVector[myID].Path.push_back(myID);
			}
		}
		void bgp_update(){
			bool ChangedInThisLoop;
			u_s TmpPathLength;
			u_s MarkedID;
			for( u_s i = 0; i < ValidNodeNum; i ++){
				ChangedInThisLoop = 0;
				if(find(AdjList.begin(), AdjList.end(), i) != AdjList.end())	
						continue;
				if(!OldPathVector[i].Validity)
						TmpPathLength = NodeNum + NodeNum;
				else
						TmpPathLength = OldPathVector[i].Path.size();
				for(list<u_s>::iterator Adj=AdjList.begin(); Adj!=AdjList.end(); Adj++){
						if(!AllNodeVector[*Adj]->OldPathVector[i].Validity)
								continue;
						if(find(AllNodeVector[*Adj]->OldPathVector[i].Path.begin(), AllNodeVector[*Adj]->OldPathVector[i].Path.end(),ID) != AllNodeVector[*Adj]->OldPathVector[i].Path.end())
								continue;
						if( AllNodeVector[*Adj]->OldPathVector[i].Path.size() + 1 < TmpPathLength){
								TmpPathLength = AllNodeVector[*Adj]->OldPathVector[i].Path.size() + 1;
								MarkedID = *Adj;
								ChangedInThisLoop = 1;
						}
				}
				if(ChangedInThisLoop){
						NewPathVector[i].Validity = 1;
						NewPathVector[i].Path = AllNodeVector[MarkedID]->OldPathVector[i].Path;
						NewPathVector[i].Path.push_front(ID);

						IsChanged = 1;
				}
			}	
		}
};

class SDN{
	public:
		u_s Size;
		vector<u_s> AsVector;
		SDN(u_s SdnSize){
			if(!SdnSize)
				return;
			if(SdnSize > ValidNodeNum)
				return;
			Size = 0;
			u_s CurrentNodeID;
			AsVector.assign(SdnSize,-1); 
			list<u_s> Queue;
			u_s StartNodeID = (u_s)(ValidNodeNum * (rand()%100/100.0));
			Queue.push_back(StartNodeID);
			AsVector[Size] = StartNodeID;
			Size ++;
			AllNodeVector[StartNodeID]->IsInSDN = 1;
			while(Size < SdnSize){	
				CurrentNodeID = Queue.front();
				Queue.pop_front();
				for(list<u_s>::iterator itor = AllNodeVector[CurrentNodeID]->AdjList.begin(); itor != AllNodeVector[CurrentNodeID]->AdjList.end(); itor ++){
					if(AllNodeVector[*itor]->IsInSDN)
						continue;
					AsVector[Size] = *itor;
					AllNodeVector[*itor]->IsInSDN = 1;
					Size ++;	
					Queue.push_back(*itor);
					if(Size == SdnSize)
						break;
				}
			}
#ifdef DEBUG
			fprintf(stderr,"=SDN List:");
			for( u_s i = 0; i < Size; i ++){
				fprintf(stderr," %d-", AsVector[i]);
			}
			fprintf(stderr,"\n");
#endif
		}
		void local_converge(){
			bool TmpIsChanged = IsChanged;
			while(1){
				IsChanged = 0;
				for(u_s i = 0; i < Size; i ++)
					AllNodeVector[AsVector[i]]->OldPathVector = AllNodeVector[AsVector[i]]->NewPathVector;
				for(u_s i = 0; i < Size; i ++){
					AllNodeVector[AsVector[i]]->bgp_update();
				}
				if(!IsChanged)
					break;
			}
			IsChanged = TmpIsChanged;
		}
};


int main(int argc, char * argv[]){
	if(argc < 5){
		fprintf(stderr,"error! not enough args\n");
		exit(-1);
	}
	NodeNum = atoi(argv[1]);
	SDNRatio = atoi(argv[2])/1000.0;
	RandSeed = atoi(argv[3]);
	RepeatTimes = atoi(argv[4]);

	ValidNodeNum = NodeNum - RepeatTimes;
	srand(RandSeed);
	AllNodeVector.assign(NodeNum, NULL);
	bool MySeed[5][5] = {{0,1,0,0,1},
						{1,0,0,1,0},
						{0,0,0,1,0},
						{0,1,1,0,0},
						{1,0,0,0,0}	};
	AdjMatrix = new bool * [NodeNum];
	for( int i = 0; i < NodeNum; i ++){
		AdjMatrix[i] = new bool [NodeNum];
		memset(AdjMatrix[i],0,NodeNum*sizeof(bool));
	}
	SFENG(AdjMatrix,MySeed);
#ifdef DEBUG
	for(u_s i = 0; i < NodeNum; i ++){
		for( u_s j = 0; j < NodeNum; j++)
			fprintf(stderr,"%d ", AdjMatrix[i][j]);
		fprintf(stderr,"\n");
	}
#endif
	for( u_s i = 0; i < ValidNodeNum; i ++){
		AllNodeVector[i] = new node(i);
		for( u_s j = 0; j < ValidNodeNum; j ++){
			if(AdjMatrix[i][j])
				AllNodeVector[i]->add_adj(j);
		}
	}
	fprintf(stderr,"initial converge time: %d\n",converge_process(0));
	MySDN = new SDN((u_s)(ValidNodeNum * SDNRatio));
	for( u_s i = 0; i < RepeatTimes; i ++){
		inject_up(ValidNodeNum);
		fprintf(stderr,"update converge time: %d\n",converge_process(1));
	}
#ifdef DEBUG
	myprint_path();
#endif
	return 0;
}
void inject_up(u_s p){
	AllNodeVector[p] = new node(p);
	for( u_s i = 0; i < p; i ++){
		if(AdjMatrix[i][p]){
			AllNodeVector[p]->add_adj(i);
			AllNodeVector[i]->add_adj(p);
			fprintf(stderr, "#	R# link up info injected: %d - %d\n",p,i); 
			break;
		}
	}
	ValidNodeNum ++;
}

u_s converge_process(bool sign){
	int LoopTimes = 0;
	while(1){
		IsChanged = 0;
		for(u_s i=0; i < ValidNodeNum; i++){
			AllNodeVector[i]->OldPathVector = AllNodeVector[i]->NewPathVector;
		}
		for(u_s i=0; i < ValidNodeNum; i++)
			AllNodeVector[i]->bgp_update();
		if(sign)
			MySDN->local_converge();	
		LoopTimes = LoopTimes + 1;
		if(!IsChanged)
			break;
	}
	return LoopTimes;
}

void myprint_path(){
	for( u_s i = 0; i < NodeNum; i ++){
		fprintf(stderr, "%d ######################- \n", i);
		for( u_s j = 0; j < NodeNum; j ++){
			fprintf(stderr,"(%d): ", j);
			if(! AllNodeVector[i]->NewPathVector[j].Validity)
				fprintf(stderr,"Invalidate\n");	
			else{
				for(list<u_s>::iterator itor = AllNodeVector[i]->NewPathVector[j].Path.begin(); itor != AllNodeVector[i]->NewPathVector[j].Path.end(); itor++)
					fprintf(stderr,"%3d ", *itor);
				fprintf(stderr,"\n");
			}
		}
	}
}

void SFENG(bool ** AdjMatrix, bool MySeed[5][5]){
	int sumlinks = 0;
	for(int i = 0; i < 5; i ++)
		for(int j = 0; j < 5; j++){
			AdjMatrix[i][j] = MySeed[i][j];
			sumlinks += MySeed[i][j];
		}
	int pos = 5;
	int rnode = 0, deg = 0;
	int linkage;
	float rvalue, rlink; 
	int mlinks = 1;
	while(pos < NodeNum){
		linkage = 0;
		rvalue = rand()%100/100.0;
		if(rvalue < 0.8)
			mlinks = 1;
		else if(rvalue < 0.95)
			mlinks = 2;
		else 
			mlinks = 3;
		while(linkage<mlinks){
			rnode = (int)(rand()%100/100.0*pos);
			deg = 0;
			for( int i = 0; i < pos; i ++){
			   	deg += AdjMatrix[rnode][i];
			}
			deg *= 2;
			rlink = rand()%100/100.0;
			if(rlink < float(deg)/sumlinks && AdjMatrix[rnode][pos]!=1){
				AdjMatrix[pos][rnode] = 1;
				AdjMatrix[rnode][pos] = 1;
				linkage = linkage + 1;
				sumlinks = sumlinks + 2;
			}
		}
		pos ++;
	}
}	
